### ISSUE TITLE
Streaming gRPC server

### ISSUE THREAD

**Issue by kyessenov** [envoyproxy/go-control-plane#6]
At the basic level, the management server provides a streaming gRPC server that implements xDS APIs. This server accepts connections from Envoys, and streams configuration updates to all proxies.

The server has to keep track of `(version_info, nonce)` for each connected `Node`, as well as the type of the requested resource (routes, listeners, clusters, or multiplexed through a single ADS stream). A dedicated go-routine will be used for each tuple `(Node, resource_type)`. These go-routines can function in parallel. Each go-routine maintains `(version_info, nonce)` state through request and response cycles.

The job of each routine is to drive the desired configuration in the remote proxies. The routine uses version and nonce to determine whether the configuration is successfully applied. If it's not applied, the routine attempts to apply it with retries.

The desired configuration is an input to each routine or the pool of routines. This can be an individual resource instance (EDS, CDS, RDS, or LDS), or a batch of resources to permit staging coordinated updates (_TBD_ in another issue).

As a first cut, we can stand up a server that hosts a pool of go-routines and manages them (by spinning out on connection request, allocating state for each, and establishing channels for communication).

cc @rshriram @junr03

**Comment by mattklein123** [comment]
> The desired configuration is an input to each routine or the pool of routines. For configuration can be an individual resource instance (EDS, CDS, RDS, or LDS), or a batch of resources to permit staging coordinated updates (TBD in another issue).

I think this part needs a bit more fleshing out. All go routines are going to effectively wait on a channel for updates from some other process (cache invalidator, etc.) that is accepting new configurations. So there should probably be some type of config watching interface that the go routine waits on. Maybe this is what you mean?

**Comment by kyessenov** [comment]
Right, the routine receives the updates and then drives its proxy to accept it. We can use a channel as an interface between the routine and the process that decides the config updates - the cache or however we call it. To layer the system, I think batching/sequencing xDSes is something in the control of the cache. The routine should just sequence the versions/nonces.

**Comment by mattklein123** [comment]
Sure +1 sounds like a good high level plan. Let's see what comes together.

**Comment by kyessenov** [comment]
Mostly done.
